fedimint-atomic-broadcast/src/network.rs

```diff
@@ -0,0 +1,98 @@
+use std::io::Write;
+
+
+use bitcoin_hashes::{sha256, Hash};
+use parity_scale_codec::{Decode, Encode, IoReader};
+
+use crate::conversion::to_peer_id;
+use crate::data_provider::UnitData;
+use crate::keychain::Keychain;
+use crate::{Message, Recipient};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Hasher;
+
+impl aleph_bft::Hasher for Hasher {
+    type Hash = [u8; 32];
+
+    fn hash(input: &[u8]) -> Self::Hash {
+        let mut engine = sha256::HashEngine::default();
+        engine
+            .write_all(input)
+            .expect("Writing to a hash engine cannot fail");
+
+        sha256::Hash::from_engine(engine).to_byte_array()
+    }
+}
+
+pub type NetworkData = aleph_bft::NetworkData<
+    Hasher,
+    UnitData,
+    <Keychain as aleph_bft::Keychain>::Signature,
+    <Keychain as aleph_bft::MultiKeychain>::PartialMultisignature,
+>;
+
+pub struct Network {
+    network_data_receiver: async_channel::Receiver<Vec<u8>>,
+    outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+}
+
+impl Network {
+    pub fn new(
+        network_data_receiver: async_channel::Receiver<Vec<u8>>,
+        outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+    ) -> Self {
+        Self {
+            network_data_receiver,
+            outgoing_message_sender,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl aleph_bft::Network<NetworkData> for Network {
+    fn send(&self, network_data: NetworkData, recipient: aleph_bft::Recipient) {
+        // convert from aleph_bft::Recipient to session::Recipient
+        let recipient = match recipient {
+            aleph_bft::Recipient::Node(node_index) => Recipient::Peer(to_peer_id(node_index)),
+            aleph_bft::Recipient::Everyone => Recipient::Everyone,
+        };
+
+        // since NetworkData does not implement Encodable we use
+        // parity_scale_codec::Encode to serialize it such that Message can
+        // implement Encodable
+        self.outgoing_message_sender
+            .try_send((Message::NetworkData(network_data.encode()), recipient))
+            .ok();
+    }
+
+    async fn next_event(&mut self) -> Option<NetworkData> {
+        // This limits the RAM consumption of a Unit to roughly 12kB
+        const ITEM_LIMIT: usize = 100;
```

Comment:
Didn't I see these constants before? Should they be one and the same constant?

Reply:
Yes in data_provider.rs. However, as they only occur in two different places I opted for duplication instead of an import.